src/sys/petsclib.cxx

```diff
@@ -47,5 +78,42 @@ void PetscLib::cleanup() {
   count = 0; // ensure that finalise is not called again later
 }
 
+void PetscLib::setPetscOptions(Options& options, const std::string& prefix) {
+  // Pass all options in the section to PETSc
+  for (auto& i : options.getChildren()) {
+    if (not i.second.isValue()) {
+      throw BoutException("Found subsection %s in %s when reading PETSc options - only "
+          "values are allowed in the PETSc options, not subsections",
+          i.first.c_str(), options.str().c_str());
+    }
+    // Note, option names in the input file don't start with "-", but need to be passed
+    // to PETSc with "-" prepended
+    PetscErrorCode ierr;
+    auto petsc_option_name = "-"+prefix+i.first;
+    if (lowercase(i.second) == "true") {
```

Comment:
If `i.second == "false"`, this goes down the other branch, which is likely not what you intended. But from what I can tell, this branch isn't actually needed, it's just that passing `nullptr` is the same as passing `true`.

Reply:
It was intended, but may well be a bad idea. I'm not sure the best way to handle this.
1. BOUT++ does handle an option with no value in `BOUT.inp` - it gives it the value `'true'` (I've made a mistake in the manual entry).
2. PETSc has flags that can be passed with no value, which should be set by passing `nullptr` as the last argument (the value) in `PetscOptionsSetValue()`.
3. I can't find an obvious value that can be passed instead of `nullptr`, e.g. passing any of `0`, `1`, `true`, `false`, `t`, `f`, `null`, `default` to `-ksp_monitor` results in a crash (apparently if there is an argument to `-ksp_monitor` it is taken to be a 'viewer' - `ascii` is valid).

There might be PETSc options that need to take the string `true` as an argument, but I haven't come across one yet. Maybe some kind of explicit null argument, e.g. `petsc_null`, would be safer, but taking `"true"` to mean 'pass no argument' lets us have input files like
```
[petsc]
ksp_monitor
options_view
```
work in the way you'd expect. I'd be in favour of leaving this code as it is, and updating the manual to warn that if you need to actually pass `"true"`, you'll need to pass it on the command line. (Alternatively, as a work-around, we could provide some argument like `petsc_true` and convert it to `true` before passing to PETSc?)

Reply:
Right! Ok, this looks like the best way of handling it then. I had taken that when the PETSc manual says "passing NULL means `true`" that all options that don't take an argument are boolean toggles, but it seems that isn't the case.

We could still maybe remove the branch here by putting the branch on the argument:

```cpp
    // If the value is "true", then this _might_ indicate a PETSc option with an
    // optional value, which is not necessarily a bool. In which case, pass `nullptr`
    const char* value = lowercase(i.second) == "true" ? nullptr : i.second.as<std::string>().c_str();

#if PETSC_VERSION_GE(3, 7, 0)
    ierr = PetscOptionsSetValue(nullptr, petsc_option_name.c_str(), value);
#else
    ierr = PetscOptionsSetValue(petsc_option_name.c_str(), value);
#endif
```

Reply:
what what? why would passing NULL mean `true` in any sensible definition?

Reply:
For clarity, it's really that you can pass `NULL` for options that don't take a value or take an optional value. In which case, it's equivalent to passing the default value, which is `"true"` for boolean flags.